src/kernel-install/kernel-install.c

```diff
@@ -0,0 +1,1177 @@
+        if (r < 0)
+                return r;
+
+        r = context_remove_entry_dir(c);
+        if (r < 0)
+                return r;
+
+        return 0;
+}
+
+static int verb_add(int argc, char *argv[], void *userdata) {
+        Context *c = ASSERT_PTR(userdata);
+        int r;
+
+        assert(argc >= 3);
+        assert(argv);
+
+        c->action = ACTION_ADD;
+
+        r = context_set_version(c, argv[1]);
+        if (r < 0)
+                return r;
+
+        r = context_set_kernel(c, argv[2]);
+        if (r < 0)
+                return r;
+
+        r = context_set_initrds(c, strv_skip(argv, 3));
+        if (r < 0)
+                return r;
+
+        r = context_prepare_execution(c);
+        if (r < 0)
+                return r;
+
+        return context_execute(c);
+}
+
+static int run_as_installkernel(int argc, char *argv[], Context *c) {
+        /* kernel's install.sh invokes us as
+         *   /sbin/installkernel <version> <vmlinuz> <map> <installation-dir>
+         * We ignore the last two arguments. */
+        if (optind + 2 > argc)
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "'installkernel' command requires at least two arguments.");
+
+        return verb_add(3, STRV_MAKE("add", argv[optind], argv[optind+1]), c);
+}
+
+static int verb_remove(int argc, char *argv[], void *userdata) {
+        Context *c = ASSERT_PTR(userdata);
+        int r;
+
+        assert(argc == 2);
+        assert(argv);
+
+        c->action = ACTION_REMOVE;
+
+        r = context_set_version(c, argv[1]);
+        if (r < 0)
+                return r;
+
+        r = context_prepare_execution(c);
+        if (r < 0)
+                return r;
+
+        return context_execute(c);
+}
+
+static int verb_inspect(int argc, char *argv[], void *userdata) {
+        Context *c = ASSERT_PTR(userdata);
+        _cleanup_free_ char *joined = NULL;
+        int r;
+
+        c->action = ACTION_INSPECT;
+
+        r = context_prepare_execution(c);
+        if (r < 0)
+                return r;
+
+        printf("%sBoot Loader Entries:%s\n", ansi_underline(), ansi_normal());
+        printf("    $BOOT: %s\n", c->boot_root);
+        printf("    Token: %s\n", c->entry_token);
+        puts("");
+
+        printf("%sUsing plugins:%s\n", ansi_underline(), ansi_normal());
+        strv_print_full(c->plugins, "    ");
+        puts("");
+
+        printf("%sPlugin environment:%s\n", ansi_underline(), ansi_normal());
+        strv_print_full(c->envp, "    ");
+        puts("");
+
+        printf("%sPlugin arguments:%s\n", ansi_underline(), ansi_normal());
+        joined = strv_join(strv_skip(c->argv, 1), " ");
+        printf("    %s\n", strna(joined));
+
+        return 0;
+}
+
+static bool bypass(void) {
+        int r;
+
+        r = getenv_bool("KERNEL_INSTALL_BYPASS");
+        if (r <= 0) {
+                if (r != -ENXIO)
+                        log_debug_errno(r, "Failed to parse $KERNEL_INSTALL_BYPASS, assuming no.");
```

Comment:
This will fail:
```console
$ SYSTEMD_LOG_LEVEL=debug KERNEL_INSTALL_BYPASS=0 build/kernel-install inspect
Assertion '(_error) != 0' failed at src/kernel-install/kernel-install.c:1087, function bypass(). Aborting.
```